### PR TITLE
Feature/ass7seltza

### PR DESCRIPTION
**Pull request by alex-g-carpenter**
assignment 7 submission

### CODE REVIEW

File: seltza/assignment7/calculator.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    pwo(n1, n2):
+        raises n1 to the n2th
+    mod(n1, n2):
+        takes n1 mod m2
+    hst():
+        prints the history, with formatting if 4 values are present
+    calculate(stin: str):
+        parses a string to perform an operation
+    """
+
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py:35]: Excellent comments

File: seltza/assignment7/calculator.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        divide two numbers
+    pwo(n1, n2):
+        raises n1 to the n2th
+    mod(n1, n2):
+        takes n1 mod m2
+    hst():
+        prints the history, with formatting if 4 values are present
+    calculate(stin: str):
+        parses a string to perform an operation
+    """
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py:34]: Excellent comments

File: seltza/assignment7/calculator.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        if n2 != 0:
+            ans = n1/n2
+            self.history = [n1, n2, ans, '/']
+            return ans
+        self.history = [n1, n2, None, '/']
+        self.razer('Error, divide by zero. Get out of here with that shit.')
+
+    def pwo(self, n1, n2):
+        """
+        adds two numbers
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py]: Does this do pow? Exponents? It's called pwo and says it adds two numbers.

File: seltza/assignment7/calculator.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        Returns:
+            ans : the ans
+        """
+        ans = pow(n1, n2)
+        self.history = [n1, n2, ans, '^']
+        return ans
+
+    def mod(self, n1, n2):
+        """
+        adds two numbers
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py]: Gottem

File: seltza/assignment7/calculator.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+        Returns:
+            ans : the ans
+        """
+        ans = n1*n2
+        self.history = [n1, n2, ans, '*']
+        return ans
+
+    def div(self, n1, n2):
+        """
+        adds two numbers
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py]: lol

File: seltza/assignment7/calculator.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+        Returns:
+            ans : the ans
+        """
+        ans = n1-n2
+        self.history = [n1, n2, ans, '-']
+        return ans
+
+    def mul(self, n1, n2):
+        """
+        adds two numbers
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py]: Lol this is mul

File: seltza/assignment7/calculator.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        Returns:
+            ans : the ans
+        """
+        ans = n1 + n2
+        self.history = [n1, n2, ans, '+']
+        return ans
+
+    def sub(self, n1, n2):
+        """
+        adds two numbers
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py]: Woopsies

File: seltza/assignment7/calculator.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    def hst(self):
+        """
+        prints the history, with formatting if 4 values are present
+
+        Returns: None
+        """
+        if self.history.__len__() == 4:
+            print('n1: ' + str(self.history[0]) + ', n2: ' + str(self.history[1]) + ', ans: ' + str(self.history[2]) + ', op: ' + self.history[3])
+        else:
+            print(self.history)
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py:168]: Cool feature

File: seltza/assignment7/calculator.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+            n2 (int or float): the first number
+
+        Returns:
+            ans : the ans
+        """
+        ans = pow(n1, n2)
+        self.history = [n1, n2, ans, '^']
+        return ans
+
+    def mod(self, n1, n2):
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py:141]: Neato functionality

File: seltza/assignment7/calculator.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+            switcher = {
+                '+': self.add,
+                '-': self.sub,
+                '*': self.mul,
+                '/': self.div,
+                '^': self.pwo,
+                '%': self.mod
+            }
+            spt = stin.split() # split string on whitespace
+            if spt.__len__() == 3: # enforce three arguments (ideally two numbers and an operator)
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py:190]: Nice error handling

File: seltza/assignment7/calculator.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    def calculate(self, stin: str):
+        """
+        parses a string to perform an operation
+
+        Parameters:
+            stin (str): string to parse
+        Returns:
+            ans : the ans
+        """
+        if stin: # protect against empty input
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py:179]: Nice error handling here too

File: seltza/assignment7/calculator.py [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+                '+': self.add,
+                '-': self.sub,
+                '*': self.mul,
+                '/': self.div,
+                '^': self.pwo,
+                '%': self.mod
+            }
+            spt = stin.split() # split string on whitespace
+            if spt.__len__() == 3: # enforce three arguments (ideally two numbers and an operator)
+                f = switcher.get(spt[1], 'nothing') # lookup the function to use from the dictionary
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py:191]: Nice inline comments

File: seltza/assignment7/calculator.py [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+                    try: # try is for ValueError raised by float() if string is malformed
+                        ans = f(float(spt[0]), float(spt[2]))
+                        # check if it's an integer, use that formatting if so
+                        if ans and ans.is_integer(): # protect against int(ans = None)
+                            return int(ans)
+                        return ans
+                    except ValueError:
+                        self.razer('You think you can get away with putting messed up numbers in here?')
+                else:
+                    self.razer('You think ' + spt[1] + ' is a valid operator? Are ya stupid or somethin?')
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py:202]: Yes very descriptive error message, well done

File: seltza/assignment7/calculator.py [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+                f = switcher.get(spt[1], 'nothing') # lookup the function to use from the dictionary
+                if f: # if f is None, the string did not contain a valid operator
+                    try: # try is for ValueError raised by float() if string is malformed
+                        ans = f(float(spt[0]), float(spt[2]))
+                        # check if it's an integer, use that formatting if so
+                        if ans and ans.is_integer(): # protect against int(ans = None)
+                            return int(ans)
+                        return ans
+                    except ValueError:
+                        self.razer('You think you can get away with putting messed up numbers in here?')
```
Comment (colinrlly) [inline seltza/assignment7/calculator.py:200]: Yeah give it to 'em

File: seltza/assignment7/driver.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+print('\nbad formatting:')
+calc(c, '4 +4')
+calc(c, '4+4')
+
+print("\nempty string:")
+calc(c, '')
+
+print("\nbad numbers:")
+calc(c, 'dog + shit')
+calc(c, '420 + 6ix9ine')
```
Comment (colinrlly) [inline seltza/assignment7/driver.py:45]: You could set up proper python testing to do this

Comment (colinrlly) [reply seltza/assignment7/driver.py:45]: Using a testing framework

**Review by colinrlly — APPROVED** [review]
V well done

File: seltza/assignment7/driver.py (hunk 15 above)
Comment (alex-g-carpenter) [reply seltza/assignment7/driver.py:45]: Yeah, I think I've only done that once. Would be a good time

Comment (colinrlly) [reply seltza/assignment7/driver.py:45]: Ye for sure. We will def test later on so you could do it here or just wait.

**Pull request merged**